File: ruby/038.rb
# @param {Integer} n
# @return {String}
def count_and_say(n)
  return "1" if n == 1
  parent_result = count_and_say(n - 1)
  parent_length = parent_result.length
  result = ""
  current = 0
  while current < parent_length do
    value = parent_result[current]
    count = 1
    while current < parent_length - 1 && parent_result[current + 1] == value do
      count += 1
      current += 1
    end

    result += (count.to_s + value)
    current += 1
  end

  result
end

File: javascript/038.js
/**
 * @param {number} n
 * @return {string}
 */
var countAndSay = function(n) {
  if (n === 1) return "1";

  strValue = countAndSay(n - 1);
  let number = "";
  let count = 0;
  let result = ""
  while(strValue.length > 0) {
    if (strValue[0] === number) {
      ++count;
    } else {
      if (count > 0) result = result + count + number;
      number = strValue[0];
      count = 1;
    }
    strValue = strValue.substring(1);
  }

  result = result + count + number;
  return result;
};